File: nuido_flow_data/static/src/components/data/update_data_node.js
// THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
//
// THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
// DO NOT USE IT IN PRODUCTION.
import { useState } from "@odoo/owl";
import { Node } from "@nuido/components/node";
import { ModelSelectorEx } from "@nuido_flow/components/ui/model_selector_ex";
import { ModelFieldSelectorEx } from "@nuido_flow/components/ui/model_field_selector_ex";
import { TextInputDialogInput } from "@nuido_base/components/dialogs/text_input_dialog_input";
export class UpdateDataNode extends Node {
    setup() {
        super.setup();
        this.state = useState({
            model: this.props.node.model,
            modelDescription: this.props.node.model_description,
            fieldName: "",
            fieldValue: ""
        });
    }
    onIdsChanged(value) {
        this.props.node.ids = value;
    }
    resetFieldState() {
        this.state.fieldName = "";
        this.state.fieldValue = "";
    }
    onModelSelected(model) {
        const { label, technical: value } = model;
        this.props.node.fields = [];
        this.state.model = value;
        this.state.modelDescription = label;
        this.props.node.model = value;
        this.props.node.model_description = label;
    }
    async onModelFieldUpdate(path, info) {
        this.state.fieldName = info.fieldDef ? info.fieldDef.name : "";
        this.refreshEdges();
    }
    onAddField() {
        if (this.state.fieldName !== "" && this.state.fieldValue !== "") {
            this.props.node.addOrUpdateField(this.state.fieldName, this.state.fieldValue);
            this.refreshEdges();
            this.resetFieldState();
        }
    }
    onRemoveField(name) {
        this.props.node.removeField(name);
        this.refreshEdges();
    }
    get modelSelectorId() {
        return `input-${this.props.node.id}}-model-selector`;
    }
    get fieldValueId() {
        return `input-${this.props.node.id}}-field-value`;
    }
}
UpdateDataNode.template = "nuido_flow_data.update-data-node";
UpdateDataNode.components = {
    ...Node.components,
    ModelSelectorEx,
    ModelFieldSelectorEx,
    TextInputDialogInput
};

File: nuido_flow_data/flows/data/update_data_node.py
# THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
#
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT
#
# THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
# DO NOT USE IT IN PRODUCTION.

import logging

from typing import Sequence

from odoo.tools import safe_eval

from odoo.addons.nuido_flow.flows.core.base_node import BaseNode
from odoo.addons.nuido_flow.flows.tools.tools import get_default_context_for_eval, get_active_record_info

from .tools import get_lookup_nodes

from odoo.addons.nuido_flow.flows.tools.log_const import LOGGER_NAME
_logger = logging.getLogger(LOGGER_NAME)

class UpdateDataNode(BaseNode):
    def _process(self, params):
        super()._process(params)

        context = get_default_context_for_eval(self.env)
        if params is not None:
            context['params'] = params

        info = get_active_record_info(self.env)
        context = {**context, **info}

        nodes, lookup_functions = get_lookup_nodes(self)

        context["nodes"] = nodes
        context["lookup"] = lookup_functions

        ids_def = self.definition["ids"]
        try:
            ids = safe_eval.safe_eval(ids_def, context)
        except:
            _logger.warning("Exception on evaluation: %s", ids_def, exc_info=True)
            ids = -1

        if not isinstance(ids, Sequence):
            ids = [ids]

        model = self.definition["model"]
        res = self.env[model]
        count = res.search_count([("id", "in", ids)])
        if count > 0:
            values = {}
            for field in self.definition["fields"]:
                try:
                    value = safe_eval.safe_eval(field["value"], context)
                except:
                    _logger.warning("Exception on evaluation: %s", field["value"], exc_info=True)
                    value = None

                field = self.env["ir.model.fields"]._get(model, field["name"])
                if field.ttype in ('one2many', 'many2many'):
                    if type(value) == list:
                        values.update({ field["name"]: value })
                    else:
                        values.update({ field["name"]: [value] })
                else:
                    values.update({ field["name"]: value })

            res = self.env[model].browse(ids)
            if (len(res) > 0 and res[0].id > -1):
                for rec in res:
                    rec.write(values)

                return res

        return {
            "result": res
        }